x-pack/plugin/analytics/src/main/java/org/elasticsearch/xpack/analytics/mapper/HistogramFieldMapper.java

```diff
@@ -439,4 +403,50 @@ protected void doXContentBody(XContentBuilder builder, boolean includeDefaults,
             builder.field(Names.IGNORE_MALFORMED, ignoreMalformed.value());
         }
     }
+
+    /** re-usable {@link HistogramValue} implementation */
+    private static class InternalHistogramValue extends HistogramValue {
+        double value;
+        int count;
+        boolean isExhausted;
+        ByteBufferStreamInput streamInput;
+
+        InternalHistogramValue() {
+        }
+
+        /** reset the value for the histogram */
+        void reset(BytesRef bytesRef) {
+            streamInput = new ByteBufferStreamInput(ByteBuffer.wrap(bytesRef.bytes, bytesRef.offset, bytesRef.length));
```

Comment:
There is little value in reusing the histogram if you still create new inputs here. You might want to have a look at `ByteArrayDataInput#reset`.

Reply:
Yes, I was a bit annoyed with that. Still now I am encoding doubles as longs and using ByteArrayDataInput for deserialising. I found a bit weird I am using different family of Input/Output classes to read / write. Is that ok / safe?

Reply:
Good question. I have a slight preference for updating the write logic to be symmetric, but could be convinced either way.

Reply:
I think this is a bit a catch 22. 

The ByteArrayDataOutput needs to be provided an array before hand so you need to know the size of the serialise object before hand. ByteBufferStreamOutput abstract out all that complexity.

The ByteBufferStreamInput is not reusable, ByteArrayDataInput is.

Maybe I am missing something but it seems something is missing. I am seeing this pattern where we are using more complex binary doc values and it seems logic to have a strategy to be reusing those wrappers, wdyt?

Reply:
I was thinking of using ByteBuffersDataOutput.

Reply:
Thanks @jpountz